File: src/test/java/com/fractalwrench/bpp/internal/ast/BppParserTest.java
package com.fractalwrench.bpp.internal.ast;

import org.junit.Test;
import org.parboiled.Rule;
import org.parboiled.parserunners.ReportingParseRunner;
import org.parboiled.support.ParsingResult;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

public class BppParserTest {

    private final BppParser bppParser = new BppParser();

    @Test
    public void testWhitespace() throws Exception {
        String [] valid = {" ", "\n", "\r", "\t"};
        String [] invalid = {"", "test"};
        checkRuleMatches(valid, invalid, bppParser.WhiteSpace());
    }

    @Test
    public void contextSwitch() throws Exception {
        fail();
    }

    @Test
    public void root() throws Exception {
        fail();
    }

    @Test
    public void block() throws Exception {
        fail();
    }

    @Test
    public void print() throws Exception {
        fail();
    }

    @Test
    public void stringLiteral() throws Exception {
        fail();
    }

    @Test
    public void getEscapeCharRules() throws Exception {
        fail();
    }

    private void checkRuleMatches(String[] valid, String invalid[], Rule rule) {
        for (String s : valid) {
            assertTrue(matchesRule(rule, s));
        }
        for (String s : invalid) {
            assertFalse(matchesRule(rule, s));
        }
    }

    private boolean matchesRule(Rule rule, String text) {
        ReportingParseRunner<Object> parseRunner = new ReportingParseRunner<>(rule);
        ParsingResult<Object> result = parseRunner.run(text);
        return result.matched;
    }

}

File: src/test/java/com/fractalwrench/bpp/internal/ast/BppParserTest.kt
package com.fractalwrench.bpp.internal.ast

import org.junit.Test
import org.parboiled.Rule
import org.parboiled.parserunners.ReportingParseRunner
import org.parboiled.support.ParsingResult

import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Assert.fail

class BppParserTest {

    private val bppParser = BppParser()

    @Test
    @Throws(Exception::class)
    fun testWhitespace() {
        val valid = arrayOf(" ", "\n", "\r", "\t")
        val invalid = arrayOf("", "test")
        checkRuleMatches(valid, invalid, bppParser.WhiteSpace())
    }

    @Test
    @Throws(Exception::class)
    fun contextSwitch() {
        fail()
    }

    @Test
    @Throws(Exception::class)
    fun root() {
        fail()
    }

    @Test
    @Throws(Exception::class)
    fun block() {
        fail()
    }

    @Test
    @Throws(Exception::class)
    fun print() {
        fail()
    }

    @Test
    @Throws(Exception::class)
    fun stringLiteral() {
        fail()
    }

    @Test
    @Throws(Exception::class)
    fun getEscapeCharRules() {
        fail()
    }

    private fun checkRuleMatches(valid: Array<String>, invalid: Array<String>, rule: Rule) {
        for (s in valid) {
            assertTrue(matchesRule(rule, s))
        }
        for (s in invalid) {
            assertFalse(matchesRule(rule, s))
        }
    }

    private fun matchesRule(rule: Rule, text: String): Boolean {
        val parseRunner = ReportingParseRunner<Any>(rule)
        val result = parseRunner.run(text)
        return result.matched
    }

}